Add input mapping to creeps tutorial

seancwall opened the pull request:
Add input mapping to the Dodge the Creeps tutorial to avoid using the `ui_*` mappings.

Addresses #5147 

See also https://github.com/godotengine/godot-demo-projects/pull/664 for corresponding demo project code change.

Screenshots showing the changed pages:

**In coding the player # 1**
<img width="753" alt="Screen Shot 2021-11-10 at 8 44 12 PM" src="https://user-images.githubusercontent.com/3537902/141222167-b914c6c4-f950-4176-9031-349587a1f2d1.png">

**In coding the player # 2**
<img width="749" alt="Screen Shot 2021-11-10 at 8 44 35 PM" src="https://user-images.githubusercontent.com/3537902/141222183-fc523320-cece-4d84-b724-62b058c1636d.png">

**In finishing up**
<img width="749" alt="Screen Shot 2021-11-10 at 8 45 18 PM" src="https://user-images.githubusercontent.com/3537902/141222198-9888921e-d75a-46c8-a764-2fd1344f41bc.png">

seancwall commented:
I wanted to point out there is some overlap in my screenshots of the proposed page changes.  Hope that's not too confusing.  😄

NathanLovato commented:
Thanks, could you please just update your images to use the same yellow outline as other images in the docs? Red is not readable for some color-blind people.

See https://docs.godotengine.org/en/latest/community/contributing/docs_writing_guidelines.html#image-contribution-guidelines

NathanLovato commented:
You'll also need to update the upstream project's code accordingly. This one's in the godot-demos repository: https://github.com/godotengine/godot-demo-projects

NathanLovato commented:
You'll also need to update the upstream project's code accordingly. This one's in the godot-demos repository: https://github.com/godotengine/godot-demo-projects

seancwall commented:
Updated the highlight colors to use 2px yellow borders:

![updated_colors](https://user-images.githubusercontent.com/3537902/141370569-614322a8-2d2f-4f29-9ca2-8018f24a2481.png)

akien-mga commented on getting_started/first_2d_game/03.coding_the_player.rst:
```diff
@@ -139,8 +139,27 @@ First, we need to check for input - is the player pressing a key? For this game,
 different keys, mouse events, or other inputs to them. For this game, we will
-just use the default events called "ui_right" etc that are assigned to the arrow
-keys on the keyboard.
+map the WASD keys to the four directions.
```
I think it's better to stick to arrow keys for this tutorial. That makes the tutorial valid for any keyboard layout, instead of only QWERTY.

akien-mga replied:
Alternatively it could use WASD with Physical Keys, but I think that's a nuance better covered in a dedicated input mapping tutorial, so here I'd keep it simple and use arrow keys.

seancwall replied:
Okay, that makes sense.  I can rework it and use the arrow keys rather than WASD.

seancwall commented:
I updated the tutorial text and screenshot to use the arrow keys rather than WASD.  I also added a note that multiple keys and/or joystick buttons/axes can be mapped to the same action:

<img width="777" alt="Screen Shot 2021-11-12 at 9 34 53 AM" src="https://user-images.githubusercontent.com/3537902/141484241-e6c850fb-358c-40b8-8ee5-76c28c0626d8.png">

NathanLovato commented on getting_started/first_2d_game/03.coding_the_player.rst:
```diff
@@ -139,8 +139,32 @@ First, we need to check for input - is the player pressing a key? For this game,
+
+.. image:: img/input-mapping-add-action.png
+
+Now we need to add a key assignment to this action. Press the "+" button and
+then press "Key" in the pop-up menu.  A dialog will ask you to press the key
+you want mapped.  Press the right arrow and then click "Ok":
```
```suggestion
We need to assign a key to this action. Click the "+" icon on the right, then
click the "Key" option in the drop-down menu. A dialog asks you to type in the
desired key. Press the right arrow on your keyboard and click "Ok".
```

seancwall replied:
Ugh, I tried to be careful about the words I used.  I'll get this right eventually.  😆

NathanLovato commented on getting_started/first_2d_game/03.coding_the_player.rst:
```diff
@@ -139,8 +139,32 @@ First, we need to check for input - is the player pressing a key? For this game,
+
+.. image:: img/input-mapping-add-key.png
+
+Repeat these steps to add three more mappings: ``move_left`` mapped to the left arrow
+key, ``move_up`` mapped to the up arrow key, and ``move_down`` mapped to the down arrow key.
+It should look something like this:
```
```suggestion
Repeat these steps to add three more mappings:

1. ``move_left`` mapped to the left arrow key.
2. ``move_up`` mapped to the up arrow key.
3. And ``move_down`` mapped to the down arrow key.

Your input map tab should look like this:
```

NathanLovato commented on getting_started/first_2d_game/03.coding_the_player.rst:
```diff
@@ -139,8 +139,32 @@ First, we need to check for input - is the player pressing a key? For this game,
+
+.. image:: img/input-mapping-completed.png
+
+Press "Close" to close the project settings.
```
```suggestion
Click the "Close" button to close the project settings.
```

NathanLovato commented on getting_started/first_2d_game/03.coding_the_player.rst:
```diff
@@ -139,8 +139,32 @@ First, we need to check for input - is the player pressing a key? For this game,
+
+Press "Close" to close the project settings.
+
+.. note::
+
+  You can map multiple keys, joystick buttons, or joystick
+  axes to the same input action.
```
```suggestion
.. note::

   We only mapped one key to each input action, but you can map multiple keys,
   joystick buttons, or mouse buttons to the same input action.
```

NathanLovato commented on getting_started/first_2d_game/03.coding_the_player.rst:
```diff
@@ -139,8 +139,32 @@ First, we need to check for input - is the player pressing a key? For this game,
-keys on the keyboard.
+map the arrow keys to the four directions.
+
+Click on *Project -> Project Settings* to open the project settings and then
+click on the *Input Map* tab at the top. Enter "move_right" and then click
+the "Add" button to add the ``move_right`` action, as shown here:
```
```suggestion
Click on *Project -> Project Settings* to open the project settings window and
click on the *Input Map* tab at the top. Type "move_right" in the top bar and
click the "Add" button to add the ``move_right`` action.
```

NathanLovato commented on getting_started/first_2d_game/07.finishing-up.rst:
```diff
@@ -45,12 +45,15 @@ Since the game is played with keyboard controls, it would be convenient if we
 could also start the game by pressing a key on the keyboard. We can do this with
 the "Shortcut" property of the ``Button`` node.
 
+First, return to the *Project -> Project Settings* and click on the *Input Map*
+tab.  Add another action (like you did for the movement keys) called
+``start_game`` mapped to ``space``.
```
I'm not fond of this one. To me, it makes sense to stick to the built-in ui_* events for UI actions. Here, we really want the `ui_select` mapping. The linked issue this PR tries to address is specifically about not using ui_* input actions for gameplay controls.

seancwall replied:
Yeah, I can see that.  You really are doing a `ui_select` type action for the shortcut.  I can remove this change from the PR.  Also, the associated demo code change PR (https://github.com/godotengine/godot-demo-projects/pull/663) can then be closed with no merge, as the demo project already uses `move_*` actions for the player movement and `ui_select` for the start button shortcut.

NathanLovato reviewed (changes requested):
I directly made modifications in the form of suggestions, which you can directly add in batch to your PR as a new commit.

 There's also one comment about the `start_game` action: I think it's fine to stick to the built-in `ui_select` action there instead.

The other comments make the writing style better match the docs' writing guidelines: explicit and precise action verbs when possible (press -> type, press -> click, etc.), I think there was one passive voice turned to the direct voice, and a couple of things like those.

seancwall commented:
Okay, I removed the changes to the "finishing up" section of the tutorial to revert back to using `ui_select` for the start button shortcut.  Assuming this is what we finally go with, I can close the associated demo project PR.

akien-mga commented:
> There's also one comment about the `start_game` action: I think it's fine to stick to the built-in `ui_select` action there instead.

Could you elaborate on why? As I pointed out in #5147, it has a different meaning that what it's being used for, and at least one user was quite confused by it in the linked Twitter thread.

NathanLovato commented:
I was thinking of `ui_accept` actually, I didn't pay attention to the action used.

What I would say is in general, it seems fine to me to use predefined actions if they match the intended input. In other words, to use the ui_* input for anything UI-related, like the main screen (even if I think it's just text and doesn't have, like, a button?). 

But whichever is fine, it's a detail at this point.

Regarding clearing up confusion, it's just a matter of mentioning there are predefined actions for UI interactions and we can use them for things like our game's main menu.

seancwall commented:
We could add some text to the tutorial where it mentions using `ui_select` (because it's already mapped to the space bar) that this is one of several actions already mapped in new projects because they are used for UI interactions, like the start game button.

NathanLovato commented:
Whichever is fine, to me, it's also good to ignore the comment I left and use a custom action. As I said, either way, I think it's a detail.

seancwall commented:
I think the current state is fine -- the player movement controls are now mapped to custom actions, which meets the intent of introducing the concept to the user. I agree that using `ui_select` versus `ui_accept` is a detail that can be ignored for now.

NathanLovato commented:
To clarify: it's using a built-in input mapping or a custom one that is a detail for me.

However, if we are going to stick to a built-in input, I think it should be `ui_accept` as the name better fits the action and also include a short explanation that this is built-in and why we use it.

Example:

> In a previous lesson, we created four input actions to move the character.
> 
> Every Godot project comes with a selection of predefined input mappings. They all start with the prefix `ui_` and can be helpful when creating user interfaces.
> 
> To start the game, we can leverage the predefined `ui_accept` action, which is mapped to the Enter key.

seancwall commented:
Made changes to the "Finishing Up" section to incorporate `ui_accept` as the start button shortcut.  Screenshot of the proposed change:

<img width="745" alt="Screen Shot 2021-11-14 at 9 28 56 AM" src="https://user-images.githubusercontent.com/3537902/141685594-f1a86359-d74b-4542-8164-1d6d354fca2d.png">

seancwall commented:
I have also created a new PR (https://github.com/godotengine/godot-demo-projects/pull/664) to make the corresponding change to the demo code.

mhilbrunner commented:
Hey! Thanks for working on this.

I would prefer if we stop relying on existing UI mappings entirely, and use a custom one like `start_game` (like suggested in https://github.com/godotengine/godot-docs/issues/5147) for consistency.

Otherwise, this looks good to merge tome :)

seancwall commented:
Per the last comment, updated the tutorial text to use a custom input action `start_game` for the start button.

mhilbrunner commented:
Thank you for contributing and for updating the PR with those changes! Merged.

akien-mga commented:
Cherry-picked for 3.4.
